### PR TITLE
Add Container Storage Interface (CSI)

### PR DESCRIPTION
**Pull request by Binyang2014**
## What changes are proposed in this pull request?

This PR integrates Alluxio CSI from community maintained repo (https://github.com/Alluxio/alluxio-csi) to main repo, with further improvement

## Why are the changes needed?

Original CSI repo is not as well curated and managed as the main repo.
Given we have more 

## Does this PR introduce any user facing changes?

Yes, see the following example

### Deploy

Go to `deploy` folder, Run following commands:
```bash
kubectl apply -f csi-alluxio-driver.yml
kubectl apply -f csi-alluxio-daemon.yml
``` 

### Example Nginx application
The `/examples` folder contains `PersistentVolume`, `PersistentVolumeClaim` and an nginx `Pod` mounting the alluxio volume under `/data`.

You will need to update the alluxio `MASTER_HOST_NAME` and the share information under `volumeAttributes` in `alluxio-pv.yaml` file to match your alluxio master configuration.

Run following commands to create nginx pod mounted with alluxio volume:
```bash
kubectl apply -f alluxio-pv.yml
kubectl apply -f alluxio-pvc.yml
kubectl apply -f nginx.yml
```

### CODE REVIEW

**Comment by Binyang2014** [comment]
List the requirements  here
- [ ] Move alluxio-csi as a module of alluxio repository
- [ ] Implements the dynamic provisioning and generate csi related yaml files and examples when using helm-generate.sh
- [ ] Using an all-in-one image in whole alluxio.

**Comment by Binyang2014** [comment]
@maobaolong Please help to review.
I am not familiar with helm chart part, may missing some file changes

File: integration/csi/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+RUN mkdir -p $GOPATH/src/github.com/Alluxio/alluxio-csi
+COPY . $GOPATH/src/github.com/Alluxio/alluxio-csi
+
+FROM dev as build
+
+RUN cd $GOPATH/src/github.com/Alluxio/alluxio-csi && \
+    CGO_ENABLED=0 go build -o /usr/local/bin/alluxio-csi
+
+FROM alluxio/alluxio-fuse:latest as final
+COPY --from=build /usr/local/bin/alluxio-csi /usr/local/bin/
```
Comment (maobaolong) [inline integration/csi/Dockerfile:25]: Please add a new line here

File: integration/csi/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# Alluxio CSI
+
+This project implement contienr storage interface(https://github.com/container-storage-interface/spec) for Alluxio.
```
Comment (maobaolong) [inline integration/csi/README.md:3]: This project -> `This module`

File: integration/csi/alluxio/controllerserver.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+func sanitizeVolumeID(volumeID string) string {
+	volumeID = strings.ToLower(volumeID)
+	if len(volumeID) > 63 {
+		h := sha1.New()
+		io.WriteString(h, volumeID)
+		volumeID = hex.EncodeToString(h.Sum(nil))
+	}
+	return volumeID
+}
```
Comment (maobaolong) [inline integration/csi/alluxio/controllerserver.go:121]: Add a new line

File: integration/csi/alluxio/driver.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+package alluxio
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/golang/glog"
+	csicommon "github.com/kubernetes-csi/drivers/pkg/csi-common"
+)
+
+const (
+	driverName = "alluxio"
```
Comment (maobaolong) [inline integration/csi/alluxio/driver.go:21]: Is this a hardcode?

File: integration/csi/alluxio/nodeserver.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+	args := []string{"mount"}
+	if len(mountOptions) > 0 {
+		args = append(args, "-o", strings.Join(mountOptions, ","))
+	}
+	args = append(args, targetPath, alluxioPath)
+	command := exec.Command("/opt/alluxio/integration/fuse/bin/alluxio-fuse", args...)
+
+	masterConfig := ""
+	if masterHost, ok := req.GetVolumeContext()["alluxio.master.hostname"]; ok {
```
Comment (maobaolong) [inline integration/csi/alluxio/nodeserver.go:76]: How to specify a `alluxio.master.hostname`?

File: integration/csi/alluxio/nodeserver.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+
+	/*
+	   https://docs.alluxio.io/os/user/edge/en/api/POSIX-API.html
+	   https://github.com/Alluxio/alluxio/blob/master/integration/fuse/bin/alluxio-fuse
+	*/
+
+	alluxioPath := req.GetVolumeContext()["alluxioPath"]
```
Comment (maobaolong) [inline integration/csi/alluxio/nodeserver.go:63]: I would be better to have the following logic

- Create an under storage space, for example ozone, s3, ceph bucket, or hdfs, cephfs folder.
- Create an alluxio mount point to the new under storage space.

When NodeUnpublishVolume, clear the new under storage space and umount the mount point.

File: integration/csi/go.mod [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+	github.com/Alluxio/alluxio-csi v0.0.0-20210426175157-bbf121c94da5
+	github.com/container-storage-interface/spec v1.1.0
+	github.com/golang/glog v0.0.0-20210429001901-424d2337a529
+	github.com/kubernetes-csi/csi-lib-utils v0.7.0 // indirect
+	github.com/kubernetes-csi/drivers v1.0.2
+	github.com/spf13/cobra v1.1.3
+	golang.org/x/net v0.0.0-20210510120150-4163338589ed
+	google.golang.org/grpc v1.37.1
+	k8s.io/mount-utils v0.21.0
+)
```
Comment (maobaolong) [inline integration/csi/go.mod:15]: please add a extra line

File: integration/csi/go.sum [hunk 8]
```diff
@@ -0,0 +1,481 @@
+cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (maobaolong) [inline integration/csi/go.sum:1]: Is this file necessary?

File: integration/csi/main.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+				cpid, _ := syscall.Wait4(-1, nil, syscall.WNOHANG, nil)
+				if cpid < 1 {
+					break
+				}
+
+				glog.V(4).Infof("Reaped process with pid %d", cpid)
+			}
+		}
+	}()
+}
```
Comment (maobaolong) [inline integration/csi/main.go:96]: add a new line

File: integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            {{- if .Values.csi.controller.resources.requests }}
+            requests:
+              cpu: {{ .Values.csi.controller.resources.requests.cpu }}
+              memory: {{ .Values.csi.controller.resources.requests.memory }}
+            {{- end }}
+          {{- end }}
+      volumes:
+        - name: socket-dir
+          emptyDir: {}
+{{- end }}
```
Comment (maobaolong) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml:109]: Add a new line

File: integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+{{ if .Values.csi.enabled -}}
+apiVersion: storage.k8s.io/v1beta1
+kind: CSIDriver
+metadata:
+  name: alluxio
+spec:
+  attachRequired: false
+  podInfoOnMount: true
+{{- end }}
```
Comment (maobaolong) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml:20]: add a new line

File: integration/kubernetes/helm-chart/alluxio/templates/csi/client-sc-nginx.yaml [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+# This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied, as more fully set forth in the License.
+#
+# See the NOTICE file distributed with this work for information regarding copyright ownership.
+#
+
+{{ if .Values.csi.clientEnabled -}}
+---
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
```
Comment (maobaolong) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/client-sc-nginx.yaml:15]: It would be better to abstract the storageClass to an individual file

**Review by maobaolong — COMMENTED** [review]
Thank you for bring alluxio-csi to alluxio repo, left some comments, please take a look.

File: integration/csi/alluxio/driver.go (hunk 4 above)
Comment (Binyang2014) [reply integration/csi/alluxio/driver.go:21]: Sure, driver name is hardcode. k8s will use this name to find correct `csi-controller` and `node-plugin`. And storage and driver should have one to one relationship.

File: integration/csi/go.sum (hunk 8 above)
Comment (Binyang2014) [reply integration/csi/go.sum:1]: It's auto generated, we should not change this file

File: integration/csi/alluxio/controllerserver.go (hunk 3 above)
Comment (Binyang2014) [reply integration/csi/alluxio/controllerserver.go:121]: Here is a new line. I think `github` hide this line.

**Comment by Binyang2014** [comment]
@maobaolong For end of line issue, if there no "\n" at the end of file, github will show: 
image

File: integration/csi/alluxio/nodeserver.go (hunk 6 above)
Comment (Binyang2014) [reply integration/csi/alluxio/nodeserver.go:63]: I think  `ozone, s3, ceph` is the under file system, only admin need to aware about this. So admin can create such folder in alluxio filesystem namespace, and mount different UFS. And provide correct storageClass such as: alluxio-ozon, alluxio-s3, alluxio-ceph.
For end user, if they want to use the storage, he/she can decide to use which storageClass or use many of them. And mount to different container path according to PVC config

File: integration/csi/alluxio/nodeserver.go (hunk 5 above)
Comment (Binyang2014) [reply integration/csi/alluxio/nodeserver.go:76]: Under `parameters` filed of storage class
such as:
```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: alluxio
provisioner: alluxio
parameters:
  alluxio.master.hostname:  master_ip
volumeBindingMode: Immediate
```
But this can only be set in `javaOptions` filed. Add this option here just for backward compatibility

**Comment by maobaolong** [comment]
@Binyang2014 Sorry, got it, thanks for share this tip.

File: integration/csi/alluxio/controllerserver.go (hunk 3 above)
Comment (maobaolong) [reply integration/csi/alluxio/controllerserver.go:121]: Sure

File: integration/csi/README.md (hunk 2 above)
Comment (Binyang2014) [reply integration/csi/README.md]: Fixed

File: integration/kubernetes/helm-chart/alluxio/templates/csi/client-sc-nginx.yaml (hunk 12 above)
Comment (Binyang2014) [reply integration/kubernetes/helm-chart/alluxio/templates/csi/client-sc-nginx.yaml]: Fixed

**Review by apc999 — APPROVED** [review]

File: integration/csi/main.go [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+func handle() {
+	startReaper()
+
+	d := alluxio.NewDriver(nodeID, endpoint)
+	d.Run()
+}
+
+/*
+Based on https://github.com/openshift/origin/blob/master/pkg/util/proc/reaper.go
```
Comment (LuQQiu) [inline integration/csi/main.go:72]: this link cannot be opened, try using a permanent link instead of the master branch link

**Review by LuQQiu — APPROVED** [review]
Left some minor comments, LGTM

**Comment by Binyang2014** [comment]
@LuQQiu @apc999.
Still has some items left.
1. Currently, we don't build the CSI image and push it to docker-hub. So customers need to build the CSI image by themselves. We need to provide CSI image or integral csi into fuse image
2. For helm chart part. May need more tests, and I want to make CSI as a deployment option. I am not sure if current implementation is correct

**Comment by maobaolong** [comment]
@Binyang2014 If we can, I hope we can maintain an all-in-one alluxio image, it include alluxio main services, fuse and csi in the future. For this PR, it looks good now.

**Comment by LuQQiu** [comment]
@Binyang2014 @maobaolong We plan to deprecate the alluxio-fuse image since the current alluxio docker image contains the fuse logics as well. 
@Binyang2014 what's needed for alluxio docker image to include the CSI logics? will it have a large impact on the docker image size? Agreed with @maobaolong that we would be better to maintain a All-in-one image if the impact is acceptable.

@jiacheliu3 can you help review the helm-chart related code in this PR? and give @Binyang2014 some suggestions for how to test them out?

**Comment by Binyang2014** [comment]
@LuQQiu We need to put CSI binary into image, it just about 13MB.

**Comment by LuQQiu** [comment]
@Binyang2014 then it's totally acceptable to put it in the docker image. Are the CSI binaries included in this PR, or how can we build the CSI binaries?

**Comment by apc999** [comment]
@jiacheliu3  can you take a look at the helmchart part?

**Comment by Binyang2014** [comment]
@LuQQiu Please refer this dockerfile  https://github.com/Binyang2014/alluxio/blob/binyli%2Fcsi/integration/csi/Dockerfile

**Comment by LuQQiu** [comment]
@Binyang2014 can you help move the CSI DockerFile logics into Alluxio DockerFile in this PR or the future PRs and update the related documentations?

**Comment by Binyang2014** [comment]
> @Binyang2014 can you help move the CSI DockerFile logics into Alluxio DockerFile in this PR or the future PRs and update the related documentations?

Doc updated. Will create an other PR to put CSi bit into alluxio Dockerfile.

File: integration/csi/README.md [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+### Deploy
+
+Please use `helm-generate.sh` to generate related templates. All CSI related templates should under `integration/kubernetes/<deploy-mode>/csi` folder.
+
+You need to deploy `alluxio-csi-controller`, `alluxio-csi-nodeplugin`, `alluxio-csi-driver` before mount volume via CSI.
+
+We provide two types of provisioning methods. For static provisioning, you need to create `PersistentVolume` and `PersistentVolumeClaim` first.
+For dynamic provisioning, you need to create `StorageClass` and  `PersistentVolumeClaim`.
+
+All these examples will be gernerated after run `helm-generate.sh`.
```
Comment (jiacheliu3) [inline integration/csi/README.md:26]: nit:
```suggestion
All these examples will be gernerated after running `helm-generate.sh`.
```

File: integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    metadata:
+      labels:
+        name: {{ $fullName }}-csi-controller
+        app: {{ $name }}
+        chart: {{ $chart }}
+        release: {{ .Release.Name }}
+        heritage: {{ .Release.Service }}
+        role: alluxio-csi-controller
+    spec:
+      hostNetwork: true
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml:46]: I'm not sure if there's a need to enable hostNetwork=false?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+      serviceAccountName: csi-controller-sa
+      containers:
+        - name: csi-provisioner
+          # For kubernetes 1.13 or above
+          image: quay.io/k8scsi/csi-provisioner:v1.4.0
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/controller.yaml:51]: If possible, I recommend having a variable here and be able to change the image

File: integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+#
+
+{{ if .Values.csi.enabled -}}
+apiVersion: storage.k8s.io/v1beta1
+kind: CSIDriver
+metadata:
+  name: alluxio
+spec:
+  attachRequired: false
+  podInfoOnMount: true
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml:19]: Not sure if these should be parameterized?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+    metadata:
+      labels:
+        name: {{ $fullName }}-csi-nodeplugin
+        app: {{ $name }}
+        chart: {{ $chart }}
+        release: {{ .Release.Name }}
+        heritage: {{ .Release.Service }}
+        role: alluxio-csi-nodeplugin
+    spec:
+      hostNetwork: true
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml:44]: Same, should we parameterize this?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+          volumeMounts:
+            - name: plugin-dir
+              mountPath: /plugin
+            - name: registration-dir
+              mountPath: /registration
+        - name: csi-nodeserver
+          securityContext:
+            privileged: true
+            runAsUser: 0
+            runAsGroup: 0
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml:83]: Is it possible to run as a non-root user?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml [hunk 20]
```diff
@@ -0,0 +1,34 @@
...
+kind: PersistentVolume
+metadata:
+  name: alluxio-pv
+  labels:
+    name: alluxio-pv
+spec:
+  accessModes:
+  - ReadWriteMany
+  capacity:
+    storage: 100Gi
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml:23]: Parameterize this?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pvc-static.yaml [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: alluxio-pvc
+spec:
+  accessModes:
+  - ReadWriteMany
+  resources:
+    requests:
+      storage: 100Gi
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/pvc-static.yaml:22]: Parameterize this?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pvc.yaml [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: alluxio-pvc
+spec:
+  accessModes:
+  - ReadWriteMany
+  resources:
+    requests:
+      storage: 100Gi
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/pvc.yaml:22]: Parameterize this?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml [hunk 23]
```diff
@@ -0,0 +1,34 @@
...
+{{ if .Values.csi.clientEnabled -}}
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: alluxio-pv
+  labels:
+    name: alluxio-pv
+spec:
+  accessModes:
+  - ReadWriteMany
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml:21]: Does it make sense for RWO too?

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pvc.yaml [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+#
+
+{{ if .Values.csi.clientEnabled -}}
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: alluxio-pvc
+spec:
+  accessModes:
+  - ReadWriteMany
```
Comment (jiacheliu3) [inline integration/kubernetes/helm-chart/alluxio/templates/csi/pvc.yaml:19]: Does it make sense for RWO too?

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]
Thanks @Binyang2014 ! Sorry about the late review. My comments are majorly about parameterizing in the helm chart you added. I think it makes sense to be able to configure a few of them. I'm open to discussions :)

File: integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml (hunk 17 above)
Comment (Binyang2014) [reply integration/kubernetes/helm-chart/alluxio/templates/csi/driver.yaml:19]: These parameters used by developer. Admin/users don't need to take care

File: integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml (hunk 19 above)
Comment (Binyang2014) [reply integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml:83]: The server need to call `mount` command which require root privilege

File: integration/csi/Dockerfile [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+# (the "License"). You may not use this work except in compliance with the License, which is
+# available at www.apache.org/licenses/LICENSE-2.0
+#
+# This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+# either express or implied, as more fully set forth in the License.
+#
+# See the NOTICE file distributed with this work for information regarding copyright ownership.
+#
+
+FROM golang:1.15.13-alpine as dev
```
[the remark this answers is not in the document]
Comment (Xenorith) [reply integration/csi/Dockerfile:12]: yes this should be fine as golang is backwards compatible

File: integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml (hunk 23 above)
Comment (Binyang2014) [reply integration/kubernetes/helm-chart/alluxio/templates/csi/pv.yaml:21]: Sure, will parameterize this. Thx

**Comment by Binyang2014** [comment]
@jiacheliu3 Thanks for review, please take a look again.

File: integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml (hunk 19 above)
Comment (jiacheliu3) [reply integration/kubernetes/helm-chart/alluxio/templates/csi/nodeplugin.yaml:83]: nit: A general recommendation is for the hard coded variables like this (where the user may want to touch in their deployments), if there's a reason to hard code (not offering an option), we should leave a comment on why this is hard-coded this way.

**Review by jiacheliu3 — APPROVED** [review]
Thanks @Binyang2014 ! Left a minor comment, LGTM. If you could resolve that minor comment that'd be even better :)

**Comment by Binyang2014** [comment]
> Thanks @Binyang2014 ! Left a minor comment, LGTM. If you could resolve that minor comment that'd be even better :)

Add comments for these fields. Thx

**Comment by LuQQiu** [comment]
Error: 9.429 [ERROR] Errors: 
Error: 9.429 [ERROR] alluxio.master.journal.raft.RaftJournalTest.gainPrimacyAfterCatchup
Error: 9.430 [ERROR]   Run 1: RaftJournalTest.gainPrimacyAfterCatchup:390->promoteFollower:404 » Timeout Tim...
Error: 9.431 [ERROR]   Run 2: RaftJournalTest.gainPrimacyAfterCatchup:390->promoteFollower:404 » Timeout Tim...
07:44:09.434 [INFO] 
Error: 9.435 [ERROR] alluxio.master.journal.raft.RaftJournalTest.gainPrimacyAfterSuspend
Error: 9.435 [ERROR]   Run 1: RaftJournalTest.gainPrimacyAfterSuspend:350->promoteFollower:404 » Timeout Tim...
Error: 9.435 [ERROR]   Run 2: RaftJournalTest.gainPrimacyAfterSuspend:350->promoteFollower:404 » Timeout Tim...

**Comment by LuQQiu** [comment]
@Binyang2014 Can you help merge master into your csi branch and update this PR? I doubt that the RaftJournalTest is already fixed in the master branch.

**Comment by apc999** [comment]
alluxio-bot, merge this please
